### PR TITLE
feat: add webhook labels flag to add labels in VPA admission controller webhook

### PR DESCRIPTION
**Pull request by umagnus**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

feat: add webhook labels flag to add labels in VPA admission controller webhook
In AKS, VPA can't admit pods in kube-system namespace. Reference to [Can admission controller webhooks impact kube-system and internal AKS namespaces?](https://learn.microsoft.com/en-us/azure/aks/faq#can-admission-controller-webhooks-impact-kube-system-and-internal-aks-namespaces-), VPA can add Label: "admissions.enforcer/disabled": "true" in admission controller webhook to impact kube-system namespace. Add a flag to add webhook labels.

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/autoscaler/issues/7392

#### Does this PR introduce a user-facing change?

```release-note
The --webhook-labels flag is added. Users can use webhook-labels flag in admission controller to add labels in vpa-webhook
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/admission-controller/main.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ var (
 	registerByURL              = flag.Bool("register-by-url", false, "If set to true, admission webhook will be registered by URL (webhookAddress:webhookPort) instead of by service name")
 	vpaObjectNamespace         = flag.String("vpa-object-namespace", apiv1.NamespaceAll, "Namespace to search for VPA objects. Empty means all namespaces will be used. Must not be used if ignored-vpa-object-namespaces is set.")
 	ignoredVpaObjectNamespaces = flag.String("ignored-vpa-object-namespaces", "", "Comma separated list of namespaces to ignore. Must not be used if vpa-object-namespace is used.")
+	webhookLabels              = flag.String("webhook-labels", "", "Comma separated list of labels to add to the webhook object. Format: key1:value1,key2:value2")
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/main.go:85]: nitpick, can this be moved higher up, near the other webhook variables?

File: vertical-pod-autoscaler/pkg/admission-controller/config.go [hunk 2]
```diff
@@ -181,3 +186,28 @@ func selfRegistration(clientset kubernetes.Interface, caCert []byte, webHookDela
 		klog.V(3).Info("Self registration as MutatingWebhook succeeded.")
 	}
 }
+
+// ConvertLabelsToMap convert the labels from string to map
+// the valid labels format is "key1:value1,key2:value2", which could be converted to
+// {"key1": "value1", "key2": "value2"}
+func ConvertLabelsToMap(labels string) (map[string]string, error) {
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/config.go:193]:
```suggestion
// convertLabelsToMap convert the labels from string to map
// the valid labels format is "key1:value1,key2:value2", which could be converted to
// {"key1": "value1", "key2": "value2"}
func convertLabelsToMap(labels string) (map[string]string, error) {
```

I think this can be private?

Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/config.go:193]: Can you add a unit test for this function?

File: vertical-pod-autoscaler/pkg/admission-controller/config.go [hunk 3]
```diff
@@ -181,3 +186,28 @@ func selfRegistration(clientset kubernetes.Interface, caCert []byte, webHookDela
...
+	}
+	s := strings.Split(labels, ",")
+	for _, tag := range s {
+		kv := strings.SplitN(tag, ":", 2)
+		if len(kv) != 2 {
+			return map[string]string{}, fmt.Errorf("labels '%s' are invalid, the format should like: 'key1:value1,key2:value2'", labels)
+		}
+		key := strings.TrimSpace(kv[0])
+		if key == "" {
+			return map[string]string{}, fmt.Errorf("labels '%s' are invalid, the format should like: 'key1:value1,key2:value2'", labels)
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/config.go:206]:
```suggestion
			return map[string]string{}, fmt.Errorf("labels '%s' are invalid, the format should be: 'key1:value1,key2:value2'", labels)
		}
		key := strings.TrimSpace(kv[0])
		if key == "" {
			return map[string]string{}, fmt.Errorf("labels '%s' are invalid, the format should be: 'key1:value1,key2:value2'", labels)
```

**Comment by adrianmoisey** [comment]
Could you also update the documentation?

File: vertical-pod-autoscaler/pkg/admission-controller/main.go (hunk 1 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go]: done

File: vertical-pod-autoscaler/pkg/admission-controller/config.go (hunk 2 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config.go]: sure, add ut and set it private

File: vertical-pod-autoscaler/pkg/admission-controller/config.go (hunk 3 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config.go]: done

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go [hunk 4]
```diff
@@ -244,3 +245,67 @@ func TestSelfRegistrationWithOutFailurePolicy(t *testing.T) {
...
+		expectedError  bool
+	}{
+		{
+			desc:           "should return empty map when tag is empty",
+			labels:         "",
+			expectedOutput: map[string]string{},
+			expectedError:  false,
+		},
+		{
+			desc:   "sing valid tag should be converted",
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/config_test.go:263]:
```suggestion
			desc:   "single valid tag should be converted",
```

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go [hunk 5]
```diff
@@ -244,3 +245,67 @@ func TestSelfRegistrationWithOutFailurePolicy(t *testing.T) {
...
+			desc:   "multiple valid labels should be converted",
+			labels: "key1:value1,key2:value2",
+			expectedOutput: map[string]string{
+				"key1": "value1",
+				"key2": "value2",
+			},
+			expectedError: false,
+		},
+		{
+			desc:   "whitespaces should be trimmed",
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/config_test.go:280]: what about adding a test with whitespace between the keys and values?
`key1 : value1,key2 : value2`

**Comment by adrianmoisey** [comment]
I'm not sure how useful it is, but what about a test for `selfRegistration()` where the input is invalid?

**Comment by adrianmoisey** [comment]
The "Does this PR introduce a user-facing change?" section of the description should be filled in

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go (hunk 4 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config_test.go]: done

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go (hunk 5 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config_test.go:324]: done

**Comment by umagnus** [comment]
> I'm not sure how useful it is, but what about a test for `selfRegistration()` where the input is invalid?

Previous I use `klog.Fatal` when label input is invalid, but I think the label may not be an error must be exited. Now I use klog.Warning to print a warning message for invalid labels and set empty labels in creating webhook. Also add the ut

**Comment by adrianmoisey** [comment]
> #### Does this PR introduce a user-facing change?
> ```
> action required: users can use webhook-labels flag in admission controller to add labels in vp
> ```

I don't think "action required" is true, since this change is backwards compatible. If a user upgrades they can ignore the flag and it shouldn't change their experience.

**Comment by umagnus** [comment]
> > #### Does this PR introduce a user-facing change?
> > ```
> > action required: users can use webhook-labels flag in admission controller to add labels in vp
> > ```
> 
> I don't think "action required" is true, since this change is backwards compatible. If a user upgrades they can ignore the flag and it shouldn't change their experience.

Sure, remove it

File: vertical-pod-autoscaler/pkg/admission-controller/config.go [hunk 6]
```diff
@@ -141,9 +141,15 @@ func selfRegistration(clientset kubernetes.Interface, caCert []byte, webHookDela
 			},
 		}
 	}
+	webhookLabelsMap, err := convertLabelsToMap(webHookLabels)
+	if err != nil {
+		klog.Warning(err)
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/config.go:146]:
```suggestion
		klog.Warningf("Unable to parse webhook labels. Creating webhook without labels: %v\n", err)
```

Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config.go]: done

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go [hunk 7]
```diff
@@ -244,3 +245,120 @@ func TestSelfRegistrationWithOutFailurePolicy(t *testing.T) {
...
+			labels:         "foo,bar",
+			expectedOutput: nil,
+			expectedError:  true,
+		},
+		{
+			desc:           "should return error for when key is missed",
+			labels:         "key1:value1,:bar",
+			expectedOutput: nil,
+			expectedError:  true,
+		},
```
[the remark this answers is not in the document]
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config_test.go:352]: Sure, I add `labels = strings.Trim(labels, "\"")` in function `convertLabelsToMap`

**Comment by voelzmo** [comment]
Hey @umagnus, thanks for the PR! Just to clarify: you actually _need_ VPA to modify Pods in `kube-system`? Are you deploying anything there, which is under control of VPA? Usually, in managed k8s distros, you want to leave that namespace alone and use your own namespaces for your components, to not interfere with whatever your provider deploys there, right?

I'm asking, because [we recently introduced a feature to _ignore_ namespaces](https://github.com/kubernetes/autoscaler/pull/6788) (other manages k8s distributions like GKE were also not happy with the VPA applying to Pods in `kube-system`), such that you can add the option `--ignored-vpa-object-namespaces=kube-system` to the vpa-recommender, vpa-updater and vpa-admission-webhook deployments. The necessary ignore matchers are automatically added to the VPA's webhook and you don't get bothered by your managed k8s distro's vendor.

**Comment by umagnus** [comment]
> Hey @umagnus, thanks for the PR! Just to clarify: you actually _need_ VPA to modify Pods in `kube-system`? Are you deploying anything there, which is under control of VPA? Usually, in managed k8s distros, you want to leave that namespace alone and use your own namespaces for your components, to not interfere with whatever your provider deploys there, right?
> 
> I'm asking, because [we recently introduced a feature to _ignore_ namespaces](https://github.com/kubernetes/autoscaler/pull/6788) (other manages k8s distributions like GKE were also not happy with the VPA applying to Pods in `kube-system`), such that you can add the option `--ignored-vpa-object-namespaces=kube-system` to the vpa-recommender, vpa-updater and vpa-admission-webhook deployments. The necessary ignore matchers are automatically added to the VPA's webhook and you don't get bothered by your managed k8s distro's vendor.

Yes, I want to use VPA to modify Pods in `kube-system` in AKS, and it need to add a label for vpa-admission-webhook deployments. AKS ignore pods in `kube-system` by default.

**Comment by umagnus** [comment]
Hi, @adrianmoisey Could you please approve it? Thanks!

**Comment by adrianmoisey** [comment]
> Hi, @adrianmoisey Could you please approve it? Thanks!

I can't approve, I can only /lgtm.

**Comment by umagnus** [comment]
Hi, @voelzmo  Could you please approve it? Thank you!

File: vertical-pod-autoscaler/pkg/admission-controller/config.go [hunk 8]
```diff
@@ -141,9 +141,15 @@ func selfRegistration(clientset kubernetes.Interface, caCert []byte, webHookDela
 			},
 		}
 	}
+	webhookLabelsMap, err := convertLabelsToMap(webHookLabels)
+	if err != nil {
+		klog.Warningf("Unable to parse webhook labels. Creating webhook without labels: %v\n", err)
```
Comment (voelzmo) [inline vertical-pod-autoscaler/pkg/admission-controller/config.go:146]: We're switching to structured logging currently (see https://github.com/kubernetes/autoscaler/pull/7226 for the admission-controller PR), therefore I'd like this to follow the same pattern. As warnings don't exist in structured logging, this probably should be logged with `klog.ErrorS`?

I'm also wondering if we should continue with the selfRegistration here or not? Currently this error just goes into the logs and everything still get registered – is this intentional?

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go [hunk 9]
```diff
@@ -244,3 +245,129 @@ func TestSelfRegistrationWithOutFailurePolicy(t *testing.T) {
...
+	timeoutSeconds := int32(32)
+	selectedNamespace := ""
+	ignoredNamespaces := []string{}
+
+	selfRegistration(testClientSet, caCert, webHookDelay, namespace, serviceName, url, registerByURL, timeoutSeconds, selectedNamespace, ignoredNamespaces, false, "foo,bar")
+
+	webhookConfigInterface := testClientSet.AdmissionregistrationV1().MutatingWebhookConfigurations()
+	webhookConfig, err := webhookConfigInterface.Get(context.TODO(), webhookConfigName, metav1.GetOptions{})
+
+	assert.NoError(t, err, "expected invalid labels error fetching webhook configuration")
```
Comment (voelzmo) [inline vertical-pod-autoscaler/pkg/admission-controller/config_test.go:267]: I'm not sure what this assertion is supposed to do. Currently, it asserts that _no error is returned_, but the message sounds like it should assert that _there will be an error returned_. What should be the expected behavior? See also my comment above about this.

**Review by voelzmo — CHANGES_REQUESTED** [review]
Thanks for the PR, just small clarification regarding logging and error behavior.

File: vertical-pod-autoscaler/pkg/admission-controller/config.go (hunk 8 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config.go]: I change it to `klog.ErrorS`. I think use labels is an option but not necessary, so when cx use invalid labels format, selfRegistration will continue and just print a error log.

File: vertical-pod-autoscaler/pkg/admission-controller/config_test.go (hunk 9 above)
Comment (umagnus) [reply vertical-pod-autoscaler/pkg/admission-controller/config_test.go]: I think webhook labels format error is an option but not necessary, so it don't need to let webhook register failed. It just need to print an error message in log and use empty labels in registering.

**Comment by voelzmo** [comment]
Thanks for the PR!

/lgtm
/approve
